astropy/io/fits/_tiled_compression/codecs.py

```diff
@@ -0,0 +1,362 @@
+"""
+This module contains the FITS compression algorithms in numcodecs style Codecs.
+"""
+from gzip import compress as gzip_compress
+from gzip import decompress as gzip_decompress
+
+import numpy as np
+
+from astropy.io.fits._tiled_compression._compression import (
+    compress_hcompress_1_c,
+    compress_plio_1_c,
+    compress_rice_1_c,
+    decompress_hcompress_1_c,
+    decompress_plio_1_c,
+    decompress_rice_1_c,
+)
+
+try:
+    from numcodecs.abc import Codec
+except ImportError:
+
+    class Codec:
+        codec_id = None
+
+
+__all__ = [
+    "Gzip1",
+    "Gzip2",
+    "Rice1",
+    "PLIO1",
+    "HCompress1",
+]
+
+
+class Gzip1(Codec):
+    """
+    The FITS GZIP 1 compression and decompression algorithm.
+
+    The Gzip algorithm is used in the free GNU software compression utility of
+    the same name. It was created by J. L. Gailly and M. Adler, based on the
+    DEFLATE algorithm (Deutsch 1996), which is a combination of LZ77 (Ziv &
+    Lempel 1977) and Huffman coding.
+    """
+
+    codec_id = "FITS_GZIP1"
+
+    def decode(self, buf):
+        """
+        Decompress buffer using the GZIP_1 algorithm.
+
+        Parameters
+        ----------
+        buf
+            The buffer to decompress.
+
+        Returns
+        -------
+        buf
+            The decompressed buffer.
+        """
+        cbytes = np.frombuffer(buf, dtype=np.uint8).tobytes()
```

Comment:
You don't need `.tobytes()` here (which makes a copy); `gzip_decompress` can handle the numpy buffer (or indeed any buffer, not sure you need even `np.frombuffer`, but at least that does not make a copy).

Reply:
Fixed

Reply:
That's very weird... Frankly sounds like some kind of bug. It may be worth trying `memoryview(buf)` or `memoryview(np.frombuffer, ...)` which would also not make a copy.

Reply:
The issue happens with compression - here's a MWE in case you are interested in following this up on the Numpy side:

```
In [1]: import numpy as np

In [2]: x = np.arange(9).reshape((3, 3)).astype(np.uint8)

In [3]: from gzip import compress

In [4]: compress(x)
Out[4]: b'\x1f\x8b\x08\x00Dk\xc6c\x02\xffc`dbfaec\xe7\x00\x00\x02C\xe1\xbc\x03\x00\x00\x00'

In [5]: compress(x.tobytes())
Out[5]: b'\x1f\x8b\x08\x00Lk\xc6c\x02\xffc`dbfaec\xe7\x00\x00\x02C\xe1\xbc\t\x00\x00\x00'

In [6]: from gzip import decompress

In [7]: decompress(compress(x))
---------------------------------------------------------------------------
BadGzipFile                               Traceback (most recent call last)
Cell In[7], line 1
----> 1 decompress(compress(x))

File /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/gzip.py:614, in decompress(data)
    612     raise BadGzipFile("CRC check failed")
    613 if length != (len(decompressed) & 0xffffffff):
--> 614     raise BadGzipFile("Incorrect length of data produced")
    615 decompressed_members.append(decompressed)
    616 data = do.unused_data[8:].lstrip(b"\x00")

BadGzipFile: Incorrect length of data produced
```

Reply:
That is very odd. It may be a `gzip` that is responsible since also `memoryview(x)`  fails.  And `zlib` works fine.

I note that in the documentation of `gzip`, it states "Changed in version 3.11: Speed is improved by compressing all data at once instead of in a streamed fashion. Calls with mtime set to 0 are delegated to [zlib.compress()](https://docs.python.org/3/library/zlib.html#zlib.compress) for better speed." (and indeed, with `mtime=0` it works)

I wonder if the "all data at once" is a clue here... But regardless, your example shows that it has something to do with strides and/or shape, and I worry that perhaps that might cause problems if the data are not contiguous. Indeed, even on python 3.10,
```
In [7]: compress(x[:, 0])
---------------------------------------------------------------------------
BufferError                               Traceback (most recent call last)
Cell In [7], line 1
----> 1 compress(x[:, 0])

File /usr/lib/python3.10/gzip.py:549, in compress(data, compresslevel, mtime)
    547 buf = io.BytesIO()
    548 with GzipFile(fileobj=buf, mode='wb', compresslevel=compresslevel, mtime=mtime) as f:
--> 549     f.write(data)
    550 return buf.getvalue()

File /usr/lib/python3.10/gzip.py:289, in GzipFile.write(self, data)
    286     length = data.nbytes
    288 if length > 0:
--> 289     self.fileobj.write(self.compress.compress(data))
    290     self.size += length
    291     self.crc = zlib.crc32(data, self.crc)

BufferError: memoryview: underlying buffer is not C-contiguous
```
So, perhaps the copy is worth it!?

For now, perhaps just add a comment...

Reply:
Added comments in https://github.com/astropy/astropy/pull/14252/commits/4690dd6505e4a432f9da704da8f8234430bc5095 (it turns out we need it for both compression and decompression)